hbw/selection/categories.py

```diff
@@ -17,6 +22,74 @@ def catid_selection_incl(self: Categorizer, events: ak.Array, **kwargs) -> tuple
+    gp_dict={},  # dict with (tuple of) pdgId + number of required prompt particles with this pdgId
+    ignore_charge=True,
+    call_force=True,
+    _operator="eq",
+)
+def catid_n_gen_particles(
+    self: Categorizer, events: ak.Array, results: SelectionResult | None = None, **kwargs,
+) -> tuple[ak.Array, ak.Array]:
+    """ Categorizer to select events with a certain number of prompt gen particles """
+
+    # possible options to compare number of particles with required number of particles: ==, >=, >, <=, <
+    assert self._operator in ("eq", "ge", "gt", "le", "lt")
+
+    # start with true mask
+    mask = np.ones(len(events), dtype=bool)
+    if self.dataset_inst.is_data:
+        # for data, always return true mask
+        return events, mask
+
+    if has_ak_column(events, "HardGenPart.pdgId"):
+        gp_id = events.HardGenPart.pdgId
+    else:
+        try:
+            # try to get gp_id column via SelectionResult
+            gp_id = events.GenPart.pdgId[results.objects.GenPart.HardGenPart]
+        except AttributeError:
+            # try to select hard gen particles via status flags
+            gp_id = events.GenPart.pdgId[events.GenPart.hasFlags("isHardProcess")]
+
+    if self.ignore_charge:
+        gp_id = abs(gp_id)
+
+    for pdgIds, required_n_particles in self.gp_dict.items():
+        # make sure that 'pdgIds' is a tuple
+        pdgIds = law.util.make_tuple(pdgIds)
+
+        # get number of gen particles with requested pdgIds for each event
+        n_particles = sum([ak.sum(gp_id == pdgId, axis=1) for pdgId in pdgIds])
+
+        # compare number of gen particles with required number of particles with requested operator
+        this_mask = getattr(n_particles, f"__{self._operator}__")(required_n_particles)
+        mask = mask & this_mask
+
+    return events, mask
+
+
+catid_gen_0lep = catid_n_gen_particles.derive("catid_gen_0lep", cls_dict={"gp_dict": {11: 0, 13: 0, 15: 0}})
+catid_gen_1e = catid_n_gen_particles.derive("catid_gen_1e", cls_dict={"gp_dict": {11: 1, 13: 0, 15: 0}})
+catid_gen_1mu = catid_n_gen_particles.derive("catid_gen_1mu", cls_dict={"gp_dict": {11: 0, 13: 1, 15: 0}})
+catid_gen_1tau = catid_n_gen_particles.derive("catid_gen_1tau", cls_dict={"gp_dict": {11: 0, 13: 0, 15: 1}})
+# catid_gen_2e = catid_n_gen_particles.derive("catid_gen_2e", cls_dict={"gp_dict": {11: 2, 13: 0, 15: 0}})
+# catid_gen_2mu = catid_n_gen_particles.derive("catid_gen_2mu", cls_dict={"gp_dict": {11: 0, 13: 2, 15: 0}})
+# catid_gen_2tau = catid_n_gen_particles.derive("catid_gen_2tau", cls_dict={"gp_dict": {11: 0, 13: 0, 15: 2}})
+# catid_gen_emu = catid_n_gen_particles.derive("catid_gen_emu", cls_dict={"gp_dict": {11: 1, 13: 1, 15: 0}})
+# catid_gen_etau = catid_n_gen_particles.derive("catid_gen_etau", cls_dict={"gp_dict": {11: 1, 13: 0, 15: 1}})
+# catid_gen_mutau = catid_n_gen_particles.derive("catid_gen_mutau", cls_dict={"gp_dict": {11: 0, 13: 1, 15: 1}})
```

Comment:
Can this be dropped?

Reply:
Yes it could, but we might want to add that for the DL channel at some point anyway